src/visitors/semantic_analysis_visitor.hpp

```diff
@@ -43,14 +46,14 @@ class SemanticAnalysisVisitor: public ConstAstVisitor {
   private:
     bool check_fail = false;
 
+    /// program symbol table
+    symtab::SymbolTable* psymtab = nullptr;
     /// true if accelerator backend is used for code generation
     bool accel_backend = false;
     /// true if the procedure or the function contains only one argument
     bool one_arg_in_procedure_function = false;
-    /// true if we are in a procedure block
-    bool in_procedure = false;
-    /// true if we are in a function block
-    bool in_function = false;
+    /// list of visited Function or Procedure Blocks
+    std::forward_list<const ast::Block*> visited_function_or_procedure_blocks;
```

Comment:
What's the reason for an `std::forward_list`?

Reply:
It's a container that allows fast access to the first element and searching in it with `std::find_if`

Reply:
Sounds like a `std::vector`. To me picking an exotic container, i.e. anything not `std::vector` reads like a lot of thought went into the choice, and even if I suspect the choice is suboptimal, I can't change it without thorough testing. The rumour is that `std::vector` is usually faster due to better cache locality and constant time random access to individual elements.

Lists have a nice property that references/pointers aren't invalidated as the container grows. Hence the question when changing this is: are parts of the code holding references, pointers to the `ast::Block*`.

Reply:
> The rumour is that std::vector is usually faster due to better cache locality and constant time random access to individual elements.

That's mostly true I believe. The only reason why I selected the `std::forward_list` is that because it might be faster than `std::vector` pushing and poping elements from its edges (to be more specific, its front). `std::vector` allows pushing and poping elements on its back but I believe that there might be cases that this will be slower than pushing and poping elements from it. I was looking to using a container that resembles a stack that is however searchable with `std::find_if`.
To increase the readability and because I believe that the performance benefit of `std::forward_list` will be insignificant compared to `std::vector` since the number of elements they hold should be really small I can revert to using an `std::vector`